### DISCUSSION TITLE
Question about installation procedure

### DISCUSSION THREAD

**Discussion by rei-vilo** [phoddie/node-red-mcu#26] [Q&A] [answered]
# Summary

On macOS, ensure Xcode and the command line tools are consistent with the macOS release.

In my case: 

* Install the tools Xcode 14.1 RC2 and CLT 14.1 RC2 for macOS 13.0

* Go for [xs-dev](https://github.com/HipsterBrown/xs-dev) to install Moddable, the underlying SDK

* The [node-red-mcu-plugin](https://github.com/ralphwetzel/node-red-mcu-plugin) installs the front-end Node-RED MCU.

---

I discovered this project at the Node-RED Con 2022. Congratulations for porting Node-RED to MCUs and bringing low-code/no-code to MCUs!

However, I am facing some issues with the installation. 

* [Moddable SDK](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/Moddable%20SDK%20-%20Getting%20Started.md), completed
* [node-red-mcu-plugin](https://github.com/ralphwetzel/node-red-mcu-plugin), completed
* butI am stuck at [Node-RED MCU Edition](https://github.com/phoddie/node-red-mcu). How to install it?

To make use even easier, is there a Docker container ready to test?

Thank you!

**Comment by ralphwetzel** [comment, marked as answer]
Hi!
Thank you for testing node-red-mcu & the plugin.
You're already ready to go: The plugin installs node-red-mcu as well...
Looking forward to your feedback!
R.

**Comment by rei-vilo** [comment]
Here is the continuation of the out-of-the-box experience...

I have applied the official fix to [SyntaxError: Unexpected token '??=' on manifest.js:286 #2 ](https://github.com/ralphwetzel/node-red-mcu-plugin), cleaner than commenting out the incriminating lines.

Node-RED shows the MCU pane. The minimal example consists of a slider as input and a text as output.

<details>
  <summary>Code</summary>
```
[
    {
        "id": "1fe04122a999db89",
        "type": "ui_switch",
        "z": "8e87e4f4f08e5f85",
        "name": "",
        "label": "switch",
        "tooltip": "",
        "group": "154df00148cb07e8",
        "order": 0,
        "width": 0,
        "height": 0,
        "passthru": true,
        "decouple": "false",
        "topic": "topic",
        "topicType": "msg",
        "style": "",
        "onvalue": "true",
        "onvalueType": "bool",
        "onicon": "",
        "oncolor": "",
        "offvalue": "false",
        "offvalueType": "bool",
        "officon": "",
        "offcolor": "",
        "animate": false,
        "className": "",
        "_mcu": {
            "mcu": true
        },
        "x": 410,
        "y": 260,
        "wires": [
            [
                "e1a767782dd7a055"
            ]
        ]
    },
    {
        "id": "e1a767782dd7a055",
        "type": "ui_text",
        "z": "8e87e4f4f08e5f85",
        "group": "154df00148cb07e8",
        "order": 1,
        "width": 0,
        "height": 0,
        "name": "",
        "label": "text",
        "format": "{{msg.payload}}",
        "layout": "row-spread",
        "className": "",
        "_mcu": {
            "mcu": true
        },
        "x": 610,
        "y": 260,
        "wires": []
    },
    {
        "id": "154df00148cb07e8",
        "type": "ui_group",
        "name": "Default",
        "tab": "7593fcf8177670fc",
        "order": 1,
        "disp": true,
        "width": "6",
        "collapse": false,
        "className": "",
        "_mcu": {
            "mcu": false
        }
    },
    {
        "id": "7593fcf8177670fc",
        "type": "ui_tab",
        "name": "Home",
        "icon": "dashboard",
        "disabled": false,
        "hidden": false,
        "_mcu": {
            "mcu": false
        }
    }
]
```
</details>

I have selected **Simulation > Moddable Two** and pressed **Build**. The console monitor displays...

> Starting Build process...
> Creating build environment for platform sim/moddable_two.

...but no more.

Are there other steps I am missing?

**Comment by ralphwetzel** [comment]
Well ... you hit three issues with one flow! 😁 

1. The plugin currently doesn't support the ui nodes. I'll add that as a [constraint](https://github.com/ralphwetzel/node-red-mcu-plugin#implemented-functionality) to the README.
2. Anticipating from the log you provided in the [issue](https://github.com/ralphwetzel/node-red-mcu-plugin/issues/2) you raised: "Node-RED MCU does not currently run on Linux "

https://github.com/phoddie/node-red-mcu/blob/5d8bde93c2b880942ed17dc990b0a5d8f6c4b87f/manifest_runtime.json#L43-L50

3. Even if it does not run on Linux (in the simulator), this should have been stated. I'll check.

**Reply by rei-vilo** [reply]
I tested the UI nodes as they were shown at the Node-RED Con 2022.

Good to know **node-red-mcu** only works on macOS.

**Reply by ralphwetzel** [reply]
> Good to know **node-red-mcu** only works on macOS.

Sorry. I guess my comment was not crisp enough:
You tried to run `node-red-mcu` in a _simulator_ on Linux. This is not supported.
Running Node-RED on Linux & building with the plugin for e.g. (a connected) ESP32 device works perfectly.

**Comment by rei-vilo** [comment]
Thank you for the clarification. 

I will try on Linux with a Raspberry Pi Pico W as target.

**Comment by rei-vilo** [comment]
Trying now against the Raspberry Pi Pico W! The local machine runs on Debian + Gnome.

I have the Pico SDK as explained at [Getting Started with Raspberry Pi Pico | Linux setup](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/pico.md#linux-setup). 

Selecting one and clicking on **Build** return this error 

> Starting Build process...
> Creating build environment for platform pico/captouch.
> Working directory: /home/reivilo/.node-red/mcu-plugin-cache/configda7refge4x
> System setup support currently not implemented for platform pico/captouch.
> cd /home/reivilo/.node-red/mcu-plugin-cache/configda7refge4x
> mcconfig -d -x localhost:5004 -m -p pico/captouch
> Error: '$(BUILD)/devices/$(PLATFORM)/targets/$(SUBPLATFORM)/host/provider': directory not found!

I have noticed all the Pico targets feature a screen. Is there an option without a screen?

Is there an example I could replicate?

---

The Node-RED minimal project used here in this example implements a basic MQTT program: received numbers are multiplied by two.

``` json
[
    {
        "id": "99367bcd13a45c26",
        "type": "mqtt in",
        "z": "8e87e4f4f08e5f85",
        "name": "",
        "topic": "TEST/IN",
        "qos": "2",
        "datatype": "auto-detect",
        "broker": "68a7573a3f55c15e",
        "nl": false,
        "rap": true,
        "rh": 0,
        "inputs": 0,
        "_mcu": {
            "mcu": false
        },
        "x": 200,
        "y": 120,
        "wires": [
            [
                "13f25af16955ad16"
            ]
        ]
    },
    {
        "id": "a9895ff8516deae2",
        "type": "mqtt out",
        "z": "8e87e4f4f08e5f85",
        "name": "",
        "topic": "TEST/OUT",
        "qos": "",
        "retain": "",
        "respTopic": "",
        "contentType": "",
        "userProps": "",
        "correl": "",
        "expiry": "",
        "broker": "68a7573a3f55c15e",
        "_mcu": {
            "mcu": false
        },
        "x": 690,
        "y": 380,
        "wires": []
    },
    {
        "id": "13f25af16955ad16",
        "type": "change",
        "z": "8e87e4f4f08e5f85",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "payload",
                "pt": "msg",
                "to": "$number(payload) * 2",
                "tot": "jsonata"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "_mcu": {
            "mcu": false
        },
        "x": 460,
        "y": 260,
        "wires": [
            [
                "a9895ff8516deae2"
            ]
        ]
    },
    {
        "id": "68a7573a3f55c15e",
        "type": "mqtt-broker",
        "name": "",
        "broker": "192.168.0.147",
        "port": "1883",
        "clientid": "",
        "autoConnect": true,
        "usetls": false,
        "protocolVersion": "4",
        "keepalive": "60",
        "cleansession": true,
        "birthTopic": "",
        "birthQos": "0",
        "birthPayload": "",
        "birthMsg": {},
        "closeTopic": "",
        "closeQos": "0",
        "closePayload": "",
        "closeMsg": {},
        "willTopic": "",
        "willQos": "0",
        "willPayload": "",
        "willMsg": {},
        "userProps": "",
        "sessionExpiry": "",
        "_mcu": {
            "mcu": false
        }
    }
]
```

**Reply by phoddie** [reply]
> I have noticed all the Pico targets feature a screen. Is there an option without a screen?

Not all the Pico targets have a screen. You can build for a bare Pico board with a platform target of `pico` and for a bare Pico W board, build with a platform target of `pico/pico_w`. A list of Pico platform targets is shown in the [table in the Platforms](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/pico.md#platforms) section of the Moddable SDK Pico documentation.

**Reply by ralphwetzel** [reply]
I've just implemented the recently added additional pico platform identifiers (e.g. `pico/pico_w`) and some small changes that should address the `directory not found!` error you hit.
Please continue to provide your feedback. This gives us / me the opportunity to ensure the full support for the pico platform.

**Comment by rei-vilo** [comment]
Thank you for your help.

Now, with board `pico/pico_w` and target `all`, compilation starts but doesn't go far...

```
Starting Build process...
Creating build environment for platform pico/pico_w.
Working directory: /home/reivilo/.node-red/mcu-plugin-cache/configda7refge4x
> cd /home/reivilo/.node-red/mcu-plugin-cache/configda7refge4x
> mcconfig -d -x localhost:5004 -m -p pico/pico_w
# library xs: xsHosts.c
# library xs: xsHost.c
# library xs: xsPlatform.c
# library xs: xsAll.c
# library xs: xsAPI.c
# library xs: xsArguments.c
# library xs: xsArray.c
# library xs: xsAtomics.c
```

However, 8 or so `arm-none-eabi-gcc` threads are running.

![Capture 2022-10-18 10-30-37](https://user-images.githubusercontent.com/951776/196379266-508295ff-ac24-4825-9fbf-d88e2c21dcf7.png)

**Comment by rei-vilo** [comment]
Maybe sharing a working container would make exploring Moddable and node-red-mcu easier.

**Comment by rei-vilo** [comment]
I tried again to install the whole project, this time on a Mac and targeting a Raspberry Pi Pico. 

Alas, the installation failed as some tools still target the Intel processors and not the Apple ARM-based processor.

Could you please share a container so I could explore Moddable and node-red-mcu?

Thank you!

**Reply by ralphwetzel** [reply]
> Alas, the installation failed as some tools still target the Intel processors and not the Apple ARM-based processor.

I already had a fix prepared for this issue. It is now [commited](https://github.com/ralphwetzel/node-red-mcu-plugin/commit/bb9851e7992b2261748c42294b8fac4e62353349).

**Reply by phoddie** [reply]
@ralphwetzel - Interesting. If that works for @rei-vilo, doesn't that suggest that the environment variable $PICO_GCC_ROOT was not set-up correct in Step 4 of the [macOS Pico set-up](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/pico.md#macos-setup) instructions? If that's the case, maybe we need to make the notice about ARM versus x86 installs more prominent / clear

FWIW – It is clever to work around problems like this in the MCU plug-in. But, ultimately it may be problematic. If the user has installed these items to a non-standard location, the MCU plug-in will override those with the standard location. That will be quite difficult to debug. I like the idea of the MCU plug-in helping users to avoid problems, but maybe there there's another approach.

**Reply by ralphwetzel** [reply]
> FWIW – It is clever to work around problems like this in the MCU plug-in. But, ultimately it may be problematic. If the user has installed these items to a non-standard location, the MCU plug-in will override those with the standard location. That will be quite difficult to debug. I like the idea of the MCU plug-in helping users to avoid problems, but maybe there there's another approach.

The plugin tries to be as supportive as possible. Environment variables are verified in two steps:
1) check if the env_var is defined (celebrate this) & points to something valid (throw if not).
2) if not, try to make a knowledgable guess & define the env_var (for the build environment) if the guess confirms.

Thus the user definition has priority & will not be overwritten. Only if there's a gap of definition, the plugin tries to help...

**Reply by ralphwetzel** [reply]
> doesn't that suggest that the environment variable $PICO_GCC_ROOT was not set-up correct in Step 4 of the [macOS Pico set-up](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/pico.md#macos-setup) instructions?

The issue could be, that those definitions do not persist the end of a terminal session.
Comparing the Pico setup instructions w/ those for ESP32, the later explicitly state to [modify the _shell startup/initialization file_](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/esp32.md#installing) (Step 7).
The Pico setup [just says](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/pico.md#macos-setup) "Set `PICO_GCC_ROOT` environment variable..." (Step 4).

Both instructions are correct & the ESP32 adds another aspect - that might be helpful for Pico (and others) as well.

**Comment by phoddie** [comment]
> I tried again to install the whole project, this time on a Mac and targeting a Raspberry Pi Pico.

Thank you for the effort. I can be frustrating when install doesn't work. Let's try to sort it out.

> Alas, the installation failed as some tools still target the Intel processors and not the Apple ARM-based processor.

Would you share the install command you used and the error message it reported?

FWIW – I regularly build for Raspberry Pi Pico on macOS using an M2 Mac and I know others using an M1 Mac. When I install, I follow the [Pico install](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/pico.md#sdk-and-host-environment-setup) instructions in the Moddable SDK. 

> Could you please share a container so I could explore Moddable and node-red-mcu?

We don't usually provide containers. We could do that, but this install should work so we'd like to sort that out. Thank you for your patience!

@mkellner do you have any suggestions here?

**Comment by rei-vilo** [comment]
Thank you for your answers and links.

## Short version

In a typical Apple way, some tools are no longer compatible with the updated releases of macOS, Xcode and Command Line Tools. Worse, compatible tools are not available except as release candidates. 

After installing Xcode 14, I had to install Xcode 14.1 RC2 and the corresponding CLT 14.1 RC2 in order to ensure compatibility with macOS 13.

## Long version

I know this is an early version of Node-RED for MCUs. However, I would like to stress that the whole installation process required me downloading ~24 GB and going through arcane commands, hardly consistent with a product which is supposed to be low code/no code. 

This is my third tentative to install Moddable and Node-RED for MCUs. Despite my interest on such a solution (see [Could Node-RED generate code for MCUs?](https://discourse.nodered.org/t/could-node-red-generate-code-for-mcus/22795) on the Node-RED forum). I have already spent too much time for installation only and I am afraid I have no more time for actually testing Moddable and Node-RED for MCUs.

Maybe a container could squeeze installation time as much as possible and let the user focus instead on the benefits of Moddable and Node-RED for MCUs. 

As a side note, I am no longer using Apple solutions for development. Two reasons: always changing hardware and software; absence of LTS policy. I learnt it the hard way with [embedXcode](https://embeddedcomputing.weebly.com/embedxcode.html), a plug-in for embedded development on Xcode I developed. I am now using Linux as a tool for professional development while I am keeping Mac as a toy.

**Reply by phoddie** [reply]
Thank you for reporting your issues and for understanding that Node-RED MCU Edition is still in its early stages. We are sorry your experience has been frustrating.  We agree it should be better. We are always open to suggestions for how to improve.

We understand that Apple tools can be challenging. To help with that and to generally eliminate the "arcane commands" from installation, Moddable is supporting a open source community-initiative called [xs-dev](https://github.com/HipsterBrown/xs-dev) that dramatically simplifies set-up of the Moddable SDK and support for MCUs.  If you can make some time and remain interested, we encourage you to contribute your expertise to help.

Finally, it may now be possible to build and run Node-RED MCU Edition on Linux. When development began, it wasn't. We'll take a look at getting that enabled so you'll be able to develop on your preferred platform.

**Comment by rei-vilo** [comment]
So now I have installed the `node-red-mcu-plugin` and added to `~/.node-red/settings.js`

```
// Please add MODDABLE *outside* the module.exports definition!
process.env.MODDABLE = "/Users/USER/Projects/moddable"
```

according to the [procedure](https://github.com/ralphwetzel/node-red-mcu-plugin#define-moddable).

Two problems:

1. Launching Node-RED throws errors...

```
% node-red
30 Oct 21:30:31 - [info] 

Welcome to Node-RED
===================

30 Oct 21:30:31 - [info] Node-RED version: v3.0.2
30 Oct 21:30:31 - [info] Node.js  version: v14.17.3
30 Oct 21:30:31 - [info] Darwin 22.1.0 x64 LE
30 Oct 21:30:31 - [info] Loading palette nodes
Error: Cannot find module 'node:child_process'
Require stack:
- /Users/USER/.node-red/node_modules/@ralphwetzel/node-red-mcu-plugin/mcu_plugin.js
- /usr/local/lib/node_modules/node-red/node_modules/@node-red/registry/lib/loader.js
- /usr/local/lib/node_modules/node-red/node_modules/@node-red/registry/lib/index.js
- /usr/local/lib/node_modules/node-red/node_modules/@node-red/runtime/lib/nodes/index.js
- /usr/local/lib/node_modules/node-red/node_modules/@node-red/runtime/lib/index.js
- /usr/local/lib/node_modules/node-red/lib/red.js
- /usr/local/lib/node_modules/node-red/red.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
    at Function.Module._load (internal/modules/cjs/loader.js:746:27)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at require (internal/modules/cjs/helpers.js:92:18)
    at Object.<anonymous> (/Users/USER/.node-red/node_modules/@ralphwetzel/node-red-mcu-plugin/mcu_plugin.js:2:18)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:14)
    at Module.require (internal/modules/cjs/loader.js:974:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/USER/.node-red/node_modules/@ralphwetzel/node-red-mcu-plugin/mcu_plugin.js',
    '/usr/local/lib/node_modules/node-red/node_modules/@node-red/registry/lib/loader.js',
    '/usr/local/lib/node_modules/node-red/node_modules/@node-red/registry/lib/index.js',
    '/usr/local/lib/node_modules/node-red/node_modules/@node-red/runtime/lib/nodes/index.js',
    '/usr/local/lib/node_modules/node-red/node_modules/@node-red/runtime/lib/index.js',
    '/usr/local/lib/node_modules/node-red/lib/red.js',
    '/usr/local/lib/node_modules/node-red/red.js'
  ]
}
30 Oct 21:30:32 - [info] Dashboard version 3.1.7 started at /ui
30 Oct 21:30:32 - [info] Settings file  : /Users/USER/.node-red/settings.js
30 Oct 21:30:32 - [info] Context store  : 'default' [module=memory]
30 Oct 21:30:32 - [info] User directory : /Users/USER/.node-red
30 Oct 21:30:32 - [info] Projects directory: /Users/USER/.node-red/projects
30 Oct 21:30:32 - [info] Server now running at http://127.0.0.1:1880/
30 Oct 21:30:32 - [info] Active project : Test_Node-RED_MCU
30 Oct 21:30:32 - [info] Flows file     : /Users/USER/.node-red/projects/Test_Node-RED_MCU/flow.json
30 Oct 21:30:32 - [warn] Using unencrypted credentials
30 Oct 21:30:32 - [info] Starting flows
30 Oct 21:30:32 - [info] Started flows
```

...and 

2. The MCU pane doesn't show up while the MCU nodes are listed.

<img width="1242" alt="Capture 2022-10-30 à 21 38 25" src="https://user-images.githubusercontent.com/951776/198900564-e75b651c-bb9b-4559-ab02-03a8c274acf0.png">

**Reply by ralphwetzel** [reply]
Thank you for reporting the results of your efforts.
The MCU pane can't show up as you hit that issue with "node:child_process". The reason for this issue: `node:child_process` was introduced w/ Node16 (what I wasn't aware of). You are working w/ Node14 - which is the recommended version to run node-RED & therefore definitely ok. I'll fix this.
Additionally: Those `mcuNodes` are not affiliated with `node-red-mcu` but another project.

**Reply by ralphwetzel** [reply]
> I'll fix this.

[Done](https://github.com/ralphwetzel/node-red-mcu-plugin/commit/3aac414533e4508d81b53d820fd6e2d82bb133af).

**Comment by rei-vilo** [comment]
Thanks! Now, that part installs and works correctly. 

Next step is to play with a minimal project a Pico. Any example to recommend?

**Reply by ralphwetzel** [reply]
> Any example to recommend?

<img width="304" alt="image" src="https://user-images.githubusercontent.com/16342003/199185571-e1223a9a-cd11-4daa-96f5-5cf7ae4c063b.png">

Despite it looks like nothing, there's happening a lot:
1) Your `button press` in the Node-RED editor is sent to Pico.
2) Pico runs the flow, creating a timestamp & forwarding it to the `DebugNode`.
3) Pico `DebugNode` logs the timestamp (output by `xsbug`) & sends it additionally back to the editor.
4) The editor displays a status message (beneath the node) & a debug message (in the debug pane).

In essence, that's the whole system in action.

```json
[
    {
        "id": "d54ee83032ee8422",
        "type": "tab",
        "label": "Flow 10",
        "disabled": false,
        "info": "",
        "env": [],
        "_mcu": {
            "mcu": true
        }
    },
    {
        "id": "4311906cded2e5d5",
        "type": "inject",
        "z": "d54ee83032ee8422",
        "name": "",
        "props": [
            {
                "p": "payload"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": "3",
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "_mcu": {
            "mcu": true
        },
        "x": 360,
        "y": 360,
        "wires": [
            [
                "359cc37ba40d4b50"
            ]
        ]
    },
    {
        "id": "359cc37ba40d4b50",
        "type": "debug",
        "z": "d54ee83032ee8422",
        "name": "Debug from MCU",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": true,
        "complete": "payload",
        "targetType": "msg",
        "statusVal": "payload",
        "statusType": "auto",
        "_mcu": {
            "mcu": true
        },
        "x": 550,
        "y": 360,
        "wires": []
    }
]
```

**Comment by rei-vilo** [comment]
Thank you @phoddie and @ralphwetzel for your answers. 

After running [xs-dev](https://github.com/HipsterBrown/xs-dev) to make sure my prior installation was fine, I managed to play with the suggested example and get the expected result.

<img width="872" alt="Capture 2022-11-01 à 13 29 20" src="https://user-images.githubusercontent.com/951776/199233221-81ba3870-9da2-44e3-9bec-334bf5ab6510.png">

<img width="1242" alt="Capture 2022-11-01 à 13 29 28" src="https://user-images.githubusercontent.com/951776/199233239-6c75f622-2e8e-47d7-a4fb-84e96a17fd8b.png">

Some comments: 

## Flows list

The MCU panes shows a long list of flows, while the current project has only one.

<img width="1242" alt="Capture 2022-11-01 à 13 29 37" src="https://user-images.githubusercontent.com/951776/199233264-96e28744-8a70-47c6-922a-b57f3df55960.png">

How to clear the list?

## JSONata

In the flow below, the payload of the MQTT message `test/in` is multiplied by two and sent back as `test/out`.

<img width="689" alt="Capture 2022-11-01 à 13 36 17" src="https://user-images.githubusercontent.com/951776/199233857-461ba951-c335-4f28-b03c-bcad1efb54e7.png">

The compiler complains against using the JSONata expression `$number(payload) * 2 + 3`.

> ### Error: cannot resolve type "jsonata"
> Error translating node ID "88b48973e67d5673, type "change", name "": Error: cannot resolve type "jsonata"

Is JSONata supported?

## WiFi credentials for Raspberry Pi Pico

Using a standard function solves the issue mentioned before.

```
msg.payload = Number(msg.payload) * 2 + 3;
return msg;
```

However, how to pass the WiFi credentials onto the Raspberry Pi Pico W?

Edit: Question about GPIO moved to #35.

To be continued...

**Reply by ralphwetzel** [reply]
> I managed to play with the suggested example and get the expected result.

🎉 


> how to pass the WiFi credentials onto the Raspberry Pi Pico W?

Click on `Show more options...` and define them as `Arguments`: 
```json
{
    "ssid": "your_ssid",
    "password": "your_password_here"
}
```

**Reply by rei-vilo** [reply]
Good to know!

Sending `2` to the MQTT broker...

```
% mosquitto_pub -h myHub.local -t test/out -m 2
```

...returns `7` as expected (2 x 2 + 3). 

```
% mosquitto_sub -h myHub.local -t test/out 
7
```

<img width="872" alt="Capture 2022-11-01 à 16 26 19" src="https://user-images.githubusercontent.com/951776/199272807-fbd21821-3b06-45bd-a987-8f3d536d5644.png">

As Node-RED is all about IoT, I would suggest a more user-friendly way of defining the WiFi credentials, for example through a configuration node.

Now waiting for an answer about the GPIO to blink the built-in Raspberry Pi Pico W LED...

**Reply by ralphwetzel** [reply]
> As Node-RED is all about IoT, I would suggest a more user-friendly way of defining the WiFi credentials, for example through a configuration node.

Feel free to contribute that! 😁

**Reply by rei-vilo** [reply]
> Feel free to contribute that! 😁

I am afraid I am just a Node-RED user!

**Reply by ralphwetzel** [reply]
> The MCU panes shows a long list of flows, while the current project has only one.

I'm pretty sure, all the other flows exist as well but are in status _hidden_. If you open the Down-Arrow menu in the top line, right hand edge of the flow tabs & select the first entry, you'll get a list of those...

![image](https://user-images.githubusercontent.com/16342003/199560617-5ba747ae-4894-4bc4-9269-9909ba4486c5.png)

**Reply by ralphwetzel** [reply]
> I am afraid I am just a Node-RED user!

Just kidding! 😉 
Based on your feedback, I've implemented an easier way to configure WiFi SSID & password:

<img width="472" alt="image" src="https://user-images.githubusercontent.com/16342003/199589469-2a87d714-faa3-497b-ab1a-c8a674882bc8.png">
